File: 1. Arrays And String/Question-5/Solution.py
#!/usr/bin/env python3

def one_edit_away(first, second):
    """
    Solution: Check if two strings are one edit (or zero edits) away
    Time Complexity: O(n) where n is the length of the shorter string
    Space Complexity: O(1) - no extra data structures used
    """
    # If lengths differ by more than 1, return False
    if abs(len(first) - len(second)) > 1:
        return False
    
    # Get shorter and longer string
    s1, s2 = (first, second) if len(first) <= len(second) else (second, first)
    
    index1 = 0  # Index for shorter string
    index2 = 0  # Index for longer string
    found_difference = False
    
    while index1 < len(s1) and index2 < len(s2):
        if s1[index1] != s2[index2]:
            # If this is the second difference, return False
            if found_difference:
                return False
            found_difference = True
            
            # If strings are same length, we're replacing a character
            # Move shorter string pointer only if strings are same length
            if len(s1) == len(s2):
                index1 += 1
        else:
            # If characters match, move shorter string pointer
            index1 += 1
        
        # Always move longer string pointer
        index2 += 1
    
    return True

def one_edit_replace(s1, s2):
    """Check if strings are one character replacement away"""
    found_difference = False
    for i in range(len(s1)):
        if s1[i] != s2[i]:
            if found_difference:
                return False
            found_difference = True
    return True

def one_edit_insert(s1, s2):
    """Check if strings are one character insertion away (s1 is shorter)"""
    index1 = 0
    index2 = 0
    while index1 < len(s1) and index2 < len(s2):
        if s1[index1] != s2[index2]:
            if index1 != index2:
                return False
            index2 += 1
        else:
            index1 += 1
            index2 += 1
    return True

def one_edit_away_alternative(first, second):
    """Alternative solution with separate functions for each edit type"""
    if len(first) == len(second):
        return one_edit_replace(first, second)
    elif len(first) + 1 == len(second):
        return one_edit_insert(first, second)
    elif len(first) == len(second) + 1:
        return one_edit_insert(second, first)
    return False

if __name__ == "__main__":
    test_cases = [
        ("pale", "ple"),     # True - remove 'a'
        ("pales", "pale"),   # True - insert 's'
        ("pale", "bale"),    # True - replace 'p' with 'b'
        ("pale", "bake"),    # False - replace 'p' and 'l'
        ("", ""),            # True - no edits
        ("", "a"),           # True - insert 'a'
        ("abc", "abcd"),     # True - insert 'd'
        ("abc", "abx"),      # True - replace 'c' with 'x'
        ("abc", "xyz")       # False - multiple changes
    ]
    
    print("Using combined approach:")
    for first, second in test_cases:
        print(f"{first}, {second}: {one_edit_away(first, second)}")
    
    print("\nUsing separate functions approach:")
    for first, second in test_cases:
        print(f"{first}, {second}: {one_edit_away_alternative(first, second)}")

File: 1. Arrays And String/Question-5/Solution.java
public class Solution {
    /**
     * Solution: Check if two strings are one edit (or zero edits) away
     * Time Complexity: O(n) where n is the length of the shorter string
     * Space Complexity: O(1) - no extra data structures used
     */
    public static boolean oneEditAway(String first, String second) {
        // If lengths differ by more than 1, return false
        if (Math.abs(first.length() - second.length()) > 1) {
            return false;
        }
        
        // Get shorter and longer string
        String s1 = first.length() <= second.length() ? first : second;
        String s2 = first.length() <= second.length() ? second : first;
        
        int index1 = 0; // Index for shorter string
        int index2 = 0; // Index for longer string
        boolean foundDifference = false;
        
        while (index1 < s1.length() && index2 < s2.length()) {
            if (s1.charAt(index1) != s2.charAt(index2)) {
                // If this is the second difference, return false
                if (foundDifference) {
                    return false;
                }
                foundDifference = true;
                
                // If strings are same length, we're replacing a character
                // Move shorter string pointer only if strings are same length
                if (s1.length() == s2.length()) {
                    index1++;
                }
            } else {
                // If characters match, move shorter string pointer
                index1++;
            }
            // Always move longer string pointer
            index2++;
        }
        
        return true;
    }
    
    /**
     * Alternative solution with separate functions for each edit type
     */
    public static boolean oneEditReplace(String s1, String s2) {
        boolean foundDifference = false;
        for (int i = 0; i < s1.length(); i++) {
            if (s1.charAt(i) != s2.charAt(i)) {
                if (foundDifference) {
                    return false;
                }
                foundDifference = true;
            }
        }
        return true;
    }
    
    public static boolean oneEditInsert(String s1, String s2) {
        // s1 is shorter, s2 is longer
        int index1 = 0;
        int index2 = 0;
        while (index1 < s1.length() && index2 < s2.length()) {
            if (s1.charAt(index1) != s2.charAt(index2)) {
                if (index1 != index2) {
                    return false;
                }
                index2++;
            } else {
                index1++;
                index2++;
            }
        }
        return true;
    }
    
    public static boolean oneEditAwayAlternative(String first, String second) {
        if (first.length() == second.length()) {
            return oneEditReplace(first, second);
        } else if (first.length() + 1 == second.length()) {
            return oneEditInsert(first, second);
        } else if (first.length() == second.length() + 1) {
            return oneEditInsert(second, first);
        }
        return false;
    }
    
    public static void main(String[] args) {
        String[][] testCases = {
            {"pale", "ple"},     // true - remove 'a'
            {"pales", "pale"},   // true - insert 's'
            {"pale", "bale"},    // true - replace 'p' with 'b'
            {"pale", "bake"},    // false - replace 'p' and 'l'
            {"", ""},            // true - no edits
            {"", "a"},           // true - insert 'a'
            {"abc", "abcd"},     // true - insert 'd'
            {"abc", "abx"},      // true - replace 'c' with 'x'
            {"abc", "xyz"}       // false - multiple changes
        };
        
        System.out.println("Using combined approach:");
        for (String[] test : testCases) {
            System.out.println(test[0] + ", " + test[1] + ": " + 
                              oneEditAway(test[0], test[1]));
        }
        
        System.out.println("\nUsing separate functions approach:");
        for (String[] test : testCases) {
            System.out.println(test[0] + ", " + test[1] + ": " + 
                              oneEditAwayAlternative(test[0], test[1]));
        }
    }
}
